hazelcast/src/main/java/com/hazelcast/map/impl/query/QueryRunner.java

```diff
@@ -68,6 +75,38 @@ public QueryRunner(MapServiceContext mapServiceContext, QueryOptimizer optimizer
         this.resultProcessorRegistry = resultProcessorRegistry;
     }
 
+    /**
+     * Runs a query on a chunk of a single partition. The chunk is defined by the offset {@code tableIndex}
+     * and the soft limit {@code fetchSize}.
+     *
+     * @param query       the query
+     * @param partitionId the partition which is queried
+     * @param tableIndex  the index at which to start querying
+     * @param fetchSize   the soft limit for the number of items to be queried
+     * @return the queryied entries along with the next {@code tableIndex} to resume querying
+     */
+    public MapEntriesWithCursor runPartitionScanQueryOnPartitionChunk(Query query, int partitionId, int tableIndex,
+                                                                      int fetchSize) {
+        final MapContainer mapContainer = mapServiceContext.getMapContainer(query.getMapName());
+        final Predicate predicate = queryOptimizer.optimize(query.getPredicate(), mapContainer.getIndexes());
+        final QueryableEntriesWithCursor entriesWithCursor =
+                partitionScanExecutor.executeOnPartitionChunk(query.getMapName(), predicate, partitionId, tableIndex, fetchSize);
+
+        final Collection<QueryableEntry> entries = entriesWithCursor.getEntries();
+
+        updateStatistics(mapContainer);
+
+        final Projection projection = query.getProjection();
+        final ArrayList<Entry<Data, Data>> projected = new ArrayList<Entry<Data, Data>>(entries.size());
+
+        for (QueryableEntry entry : entries) {
+            final Data value = serializationService.toData(projection.transform(entry));
+            projected.add(new SimpleEntry<Data, Data>(entry.getKeyData(), value));
```

Comment:
The problem with ENTRY was that the `QueryResult` didn't run a predicate and projection on that iteration type : https://github.com/hazelcast/hazelcast/blob/77ee1969666d385b1bcbc6a6294911f398b7e9fd/hazelcast/src/main/java/com/hazelcast/map/impl/query/QueryResult.java#L112
And I didn't want to change the existing code just to be safe.
If it is ok we can add projections and predicates on ENTRY as well.